### PR TITLE
fix color in draw_segmentation_masks

### PR DESCRIPTION
**Pull request by rizavelioglu**
**Reproduce bug:**
```python
import torch
import torchvision.utils as utils

num_masks, h, w = 4, 100, 100
img = torch.randint(0, 256, size=(3, h, w), dtype=torch.uint8)
masks = torch.randint(0, 2, (num_masks, h, w), dtype=torch.bool)

utils.draw_segmentation_masks(img, masks, colors="red")
```
which throws: 
> ValueError: There are more masks (4) than colors (3)

where the expected outcome is the drawing of all the segmentation masks in a single color ([as stated in the docs](https://pytorch.org/vision/main/generated/torchvision.utils.draw_segmentation_masks.html)).

Notice that, in the following settings the function works as intended;
- `num_masks=3` and `colors="red"`,
- `num_masks=4` and `colors="blue"`,
meaning, in the case of `num_masks=len(colors)`.

Currently, the respective tests pass even though the function does not work properly. This is because the test is checked in the presence of only 2 masks:
https://github.com/pytorch/vision/blob/b78d98bb152ffb9c0c0f5365f59f475c70b1784e/test/test_utils.py#L205
Therefore, for any input whose length is larger than 2, _i.e._ `"red"`, `"#FF00FF"`, `(240, 10, 157)`, the tests will pass.

**The Fix:**
I have adapted the code in `draw_segmentation_masks` such that it also accepts a string of colors. The code was adapted similarly to how it is handled in [`draw_bounding_boxes`](https://github.com/pytorch/vision/blob/b78d98bb152ffb9c0c0f5365f59f475c70b1784e/torchvision/utils.py#L220-L228), where the issue does not appear.
Specific tests passed with:
```python
pytest test/test_utils.py -vvv -k test_draw_segmentation_masks
```

### CODE REVIEW

**Review by pmeier — COMMENTED** [review]
Thanks @rizavelioglu for the PR. The idea to align with the bounding box functionality SGTM. IMO, we should have a `def _parse_colors(colors, num_objects: int)` function that handles this. With that, we don't need to duplicate the parsing in both functions. Since it would be a "generalization" of `_generate_color_palette`, we could inline that into the new function. Thoughts?

File: torchvision/utils.py [hunk 1]
```diff
@@ -307,38 +307,36 @@ def draw_segmentation_masks(
...
+    elif isinstance(colors, list):
+        if len(colors) < num_masks:
+            raise ValueError(f"Number of colors ({len(colors)}) is less than the number of masks ({num_masks}). ")
+    elif not isinstance(colors, (tuple, str)):
         raise ValueError("colors must be a tuple or a string, or a list thereof")
-    if isinstance(colors[0], tuple) and len(colors[0]) != 3:
+    elif isinstance(colors, tuple) and len(colors) != 3:
         raise ValueError("It seems that you passed a tuple of colors instead of a list of colors")
-
-    out_dtype = torch.uint8
```
[the remark this answers is not in the document]
Comment (rizavelioglu) [reply torchvision/utils.py:327]: I reverted it [in the new `_parse_colors` function](https://github.com/rizavelioglu/vision/blob/5a03b2b01a6fb00522be76e8dc97910090a73c80/torchvision/utils.py#L598).

**Comment by rizavelioglu** [comment]
Thank you @pmeier for your message!
I applied the changes and added the `_parse_colors` function. but with an additional argument: `output_format`. 
I did so because `draw_bounding_boxes` expects a list of `tuple`, whereas `draw_segmentation_masks` expects a list of `torch.tensor`. What do you think?

File: torchvision/utils.py [hunk 2]
```diff
@@ -575,3 +549,63 @@ def _make_ntuple(x: Any, n: int) -> Tuple[Any, ...]:
     if isinstance(x, collections.abc.Iterable):
         return tuple(x)
     return tuple(repeat(x, n))
+
+
+def _parse_colors(
+    colors: Optional[Union[List[Union[str, Tuple[int, int, int]]], str, Tuple[int, int, int], None]],
```
Comment (pmeier) [inline torchvision/utils.py:555]: 1. Reordering for better legibility
2. `Optional[Union[..., None]] == Union[..., None]`

```suggestion
    colors: Union[None, str, Tuple[int, int, int], List[Union[str, Tuple[int, int, int]]]],
```

File: torchvision/utils.py [hunk 3]
```diff
@@ -575,3 +549,63 @@ def _make_ntuple(x: Any, n: int) -> Tuple[Any, ...]:
...
+    num_objects: int,
```
Comment (pmeier) [inline torchvision/utils.py:556]: Nit: Let's enforce keyword only here

```suggestion
    *,
    num_objects: int,
```

File: torchvision/utils.py [hunk 4]
```diff
@@ -307,38 +299,20 @@ def draw_segmentation_masks(
...
-        raise ValueError("It seems that you passed a tuple of colors instead of a list of colors")
-
-    out_dtype = torch.uint8
-
-    colors_ = []
-    for color in colors:
-        if isinstance(color, str):
-            color = ImageColor.getrgb(color)
-        colors_.append(torch.tensor(color, dtype=out_dtype))
+    colors = _parse_colors(colors=colors, num_objects=num_masks)
```
Comment (pmeier) [inline torchvision/utils.py:307]: Nit: Function name is expressive enough to not needing the pass `colors` as keyword argument:

```suggestion
    colors = _parse_colors(colors, num_objects=num_masks)
```

**Review by pmeier — COMMENTED** [review]
Answered your question inline. Looking good so far!

File: test/test_utils.py [hunk 5]
```diff
@@ -195,8 +198,8 @@ def test_draw_no_boxes():
         "blue",
         "#FF00FF",
         (1, 34, 122),
-        ["red", "blue"],
-        ["#FF00FF", (1, 34, 122)],
+        ["red", "#FF00FF"],
+        ["blue", (1, 34, 122)],
```
[the remark this answers is not in the document]
Comment (rizavelioglu) [reply test/test_utils.py]: reverted in [f17249d](https://github.com/pytorch/vision/pull/7520/commits/f17249d019b8c71d0cefea4c81d80798b78d71d4)

File: torchvision/utils.py [hunk 6]
```diff
@@ -575,3 +549,63 @@ def _make_ntuple(x: Any, n: int) -> Tuple[Any, ...]:
...
+    Returns:
+        A list of colors. If output_format is "tensor", the list contains torch.Tensors. If output_format is "tuple",
+        the list contains tuples.
+
+    Raises:
+        ValueError: If the number of colors in the list is less than the number of objects to color.
+                    If `colors` is not one of the above types.
+    """
+    if colors is None:
+        colors = _generate_color_palette(num_objects)
```
[the remark this answers is not in the document]
Comment (rizavelioglu) [reply torchvision/utils.py]: applied the changes in [f17249d](https://github.com/pytorch/vision/pull/7520/commits/f17249d019b8c71d0cefea4c81d80798b78d71d4)

File: torchvision/utils.py [hunk 7]
```diff
@@ -575,3 +549,63 @@ def _make_ntuple(x: Any, n: int) -> Tuple[Any, ...]:
...
+            if isinstance(color, str)
+            else torch.tensor(color, dtype=out_dtype)
+            for color in colors
+        ]
+    elif output_format == "tuple":
+        colors_ = [ImageColor.getrgb(color) if isinstance(color, str) else color for color in colors]
+    else:
+        raise ValueError(f"Invalid output_format: {output_format}. Must be 'tensor' or 'tuple'.")
+
+    return colors_
```
[the remark this answers is not in the document]
Comment (rizavelioglu) [reply torchvision/utils.py]: applied the changes in [f17249d](https://github.com/pytorch/vision/pull/7520/commits/f17249d019b8c71d0cefea4c81d80798b78d71d4)

File: torchvision/utils.py [hunk 8]
```diff
@@ -575,3 +545,50 @@ def _make_ntuple(x: Any, n: int) -> Tuple[Any, ...]:
...
+    elif not isinstance(colors, (tuple, str)):
+        raise ValueError("`colors` must be a tuple or a string, or a list thereof, but got {colors}.")
+    elif isinstance(colors, tuple) and len(colors) != 3:
+        raise ValueError("If passed as tuple, colors should be an RGB triplet, but got {colors}.")
+    else:  # colors specifies a single color for all objects
+        colors = [colors] * num_objects
+
+    colors_ = [ImageColor.getrgb(color) if isinstance(color, str) else color for color in colors]
+
+    return colors_
```
Comment (pmeier) [inline torchvision/utils.py:594]:
```suggestion
    return [ImageColor.getrgb(color) if isinstance(color, str) else color for color in colors]
```

File: test/test_utils.py [hunk 9]
```diff
@@ -223,7 +226,8 @@ def test_draw_segmentation_masks(colors, alpha):
     assert_equal(img[:, ~masked_pixels], out[:, ~masked_pixels])
 
     if colors is None:
-        colors = utils._generate_color_palette(num_masks)
+        palette = torch.tensor([2**25 - 1, 2**15 - 1, 2**21 - 1])
+        colors = [tuple((i * palette) % 255) for i in range(num_masks)]
```
Comment (pmeier) [inline test/test_utils.py:230]: Hmmm, I was not aware of this usage. Maybe we should have the `_generate_color_palette` function after all? I lean towards putting it back, but no strong opinion.

Comment (rizavelioglu) [reply test/test_utils.py:230]: I agree, will put it back

File: torchvision/utils.py [hunk 10]
```diff
@@ -575,3 +545,52 @@ def _make_ntuple(x: Any, n: int) -> Tuple[Any, ...]:
...
+
+    Args:
+        colors: A specification of colors for the objects. This can be one of the following:
+            - None: to generate a color palette automatically.
+            - A list of colors: where each color is either a string (specifying a named color) or an RGB tuple.
+            - A string or an RGB tuple: to use the same color for all objects.
+
+            If `colors` is a tuple, it should be a 3-tuple specifying the RGB values of the color.
+            If `colors` is a list, it should have at least as many elements as the number of objects to color.
+            If `colors` is None, a color palette will be generated automatically to have a distinct color for each object.
```
Comment (pmeier) [inline torchvision/utils.py:571]: I feel this is duplicated from the listing above?

```suggestion
```

File: torchvision/utils.py [hunk 11]
```diff
@@ -575,3 +545,52 @@ def _make_ntuple(x: Any, n: int) -> Tuple[Any, ...]:
     if isinstance(x, collections.abc.Iterable):
         return tuple(x)
     return tuple(repeat(x, n))
+
+
+def _generate_color_palette(num_objects: int):
```
Comment (pmeier) [inline torchvision/utils.py:550]: Nit: could you put before `_log_api_usage_once` again to avoid unnecessary diff? If you like, you can also move `_parse_colors` there to "group" both functions together.

**Review by pmeier — APPROVED** [review]
Sorry for the delay @rizavelioglu. Two more minor things that I found in a final pass, but they are non-blocking. Let me know if you want to address them. Otherwise we can also merge as is.

**Comment by rizavelioglu** [comment]
Thanks @pmeier, been waiting for your reply :)
Sure, I will apply the changes.

File: torchvision/utils.py (hunk 11 above)
Comment (rizavelioglu) [reply torchvision/utils.py]: applied in [658b98d](https://github.com/pytorch/vision/pull/7520/commits/658b98d31fe49aa87c4fb1953c309fef28d58bda)
